### PR TITLE
Replace futures crate with pollster.

### LINKED ISSUE #628: Reduce `futures` dependencies
Looks like the `futures` drag a bunch of other stuff. Perhaps, we can make them optional? Or only depend on `futures-core`? or nothing at all, and implement the needed stuff directly on `std`.

### PR DESCRIPTION
**Pull request by parasyte**
This PR removes all of the `futures` dependencies. `std::future` does not contain a lot of useful helpers available in `futures`. The obvious ones are `join_all`, `FutureExt::map`, and `block_on`.

- `join_all` is replaced with a `Vec<T>` and an `async` block.
- `FuturesExt::map` in the web backend is replaced by rolling the `map` function into the `MakeSendFuture` type.
- `block_on` is provided by `pollster`.

The original code using `join_all` ignored the result type yielded by the Future from `map_async`. This code does the same, but makes dropping the result a little more obvious.

These should not be troublesome. Figured I would call them out anyway.

The last big change is replacing `futures-executor` in the examples with `async-executor`. A new concrete `Spawner` type is used in the example framework instead of an implementation of `futures_task::LocalSpawn`.

Fixes #628

### CODE REVIEW

File: src/backend/web.rs
Comment (grovesNL) [inline src/backend/web.rs:802]: Does this allocate now? Could we keep the old signature somehow?

Comment (parasyte) [reply src/backend/web.rs:802]: It does allocate now. I have to use `Box<dyn Future<...>>` to return an `async` block. Which I am only doing because it's nicer to use than creating a concrete type that implements `Future`. This is definitely something than can be improved though.

Comment (parasyte) [reply src/backend/web.rs]: @grovesNL I copy-pasted `FuturesExt::Map` to get this working. Fair tradeoff, I suppose.

**Comment by kvark** [comment]
@parasyte thank you for taking a stab at this! I'm excite to see the reduced dependencies list. This and the formely surfman dependencies were identified as the biggest sources of dependencies for us.
@grovesNL would you be able to make a final call on this? I believe you are in the best position to do so, but of course it depends on your availability.

File: src/backend/web.rs
Comment (grovesNL) [inline src/backend/web.rs:80]: Could we avoid `pin_project` here too? I think we should be able to do something simpler here like we have for `MakeSendFuture` (possibly even collapsing them together)

Comment (parasyte) [reply src/backend/web.rs:80]: We can, but `pin-project-lite` is already a dependency of `naga`, so I am not sure what we gain by making that change. Merging with `MakeSendFuture` sounds like a good idea!

Comment (kvark) [reply src/backend/web.rs:80]: I'm curious as to where Naga would need it.
Anyway, Naga is not a wgpu-rs dependency for the `wasm32` target.

Comment (parasyte) [reply src/backend/web.rs:80]: I misspoke. not Naga. But `tracing`:

```
tracing v0.1.22
├── cfg-if v1.0.0
├── pin-project-lite v0.2.0
├── tracing-attributes v0.1.11 (proc-macro)
│   ├── proc-macro2 v1.0.24
│   │   └── unicode-xid v0.2.1
│   ├── quote v1.0.7
│   │   └── proc-macro2 v1.0.24 (*)
│   └── syn v1.0.54
│       ├── proc-macro2 v1.0.24 (*)
│       ├── quote v1.0.7 (*)
│       └── unicode-xid v0.2.1
└── tracing-core v0.1.17
    └── lazy_static v1.4.0
```

Comment (grovesNL) [reply src/backend/web.rs:80]: Alright :+1: I think ideally tracing shouldn't be a dependency for wasm (at least by default), so removing the dependency would still be useful to keep wasm builds smaller. But we could probably tackle this in a follow-up PR instead

**Review by grovesNL — APPROVED** [review]

**Comment by grovesNL** [comment]
Oh I guess the only other concern is that we have (temporarily) copy-pasted code from `futures-util` – are there any licensing concerns because of that?

File: src/backend/web.rs
Comment (kvark) [reply src/backend/web.rs:80]: Agreed, we don't need to log anything with tracing when targeting the Web, but can follow-up.
Although, now, instead of just cfg-blocking `tracing` usage, we'd have to figure out how to rewrite the pin section...
So it would be nicer to deal with it right away.

**Comment by parasyte** [comment]
`futures-util` is dual-licensed under Apache 2.0 and MIT:

- https://github.com/rust-lang/futures-rs/blob/master/LICENSE-APACHE
- https://github.com/rust-lang/futures-rs/blob/master/LICENSE-MIT

I don't know how well these are compatible with MPL. If licensing is an issue, the `Map` and its supporting code will need to be replaced with I-don't-know-what. :)

**Comment by grovesNL** [comment]
APACHE 2/MIT are ok, I was just wondering how to properly attribute the source code to the original authors (e.g. copyright notice in that file or anything else we need). It might be easier to replace it in this PR anyway to avoid the copy-pasted code.

**Comment by parasyte** [comment]
@grovesNL I was able to combine `MakeSendFuture` and `Map` into a single type. This last commit also removes all of the code borrowed from the `Futures` crate.

I do need someone to thoroughly check my work here, however. I'm not entirely familiar with `Pin` and `Unpin`. I used the latter to avoid some extra unsafe blocks. AFAICT this is ok according to the docs, since it only accesses the `map` field non-structurally. I also added an assertion to panic if the future is ever polled after returning `Ready`. If any of this needs to be changed, I'll be glad to.

Last thing to mention is that I can squash merge all of these commits if you would like.

**Comment by parasyte** [comment]
Rebased and force-pushed as a squashed commit.

**Comment by grovesNL** [comment]
Thank you for the update!

I'll try to review the async changes soon – I'm taking a moment to get caught up on some of the futures changes over the past couple years. If anyone else is familiar with futures/pin and able to review it in the meantime, please go ahead :+1:

File: src/backend/web.rs
Comment (grovesNL) [inline src/backend/web.rs:59]: We probably don't need to worry about the `map` being used multiple times, because the caller should stop polling once `Poll::Ready` happens anyway.  So maybe we could remove the `Option` here to just have `M`, and also remove the `assert`

File: src/backend/web.rs
Comment (grovesNL) [inline src/backend/web.rs:95]: We shouldn't need `Unpin` here

File: src/backend/web.rs
Comment (grovesNL) [inline src/backend/web.rs:73]: We could probably simplify the access to `T` a little, e.g.:
```rust
fn poll(self: Pin<&mut Self>, cx: &mut task::Context) -> Poll<Self::Output> {
    // This is safe because we have no Drop implementation to violate the Pin requirements and
    // do not provide any means of moving the inner future.
    unsafe {
        let this = self.get_unchecked_mut();
        match Pin::new_unchecked(&mut this.future).poll(cx) {
            task::Poll::Ready(value) => task::Poll::Ready((this.map)(value)),
            task::Poll::Pending => task::Poll::Pending,
        }
    }
}
```

File: src/util/belt.rs
Comment (grovesNL) [inline src/util/belt.rs:141]: If I'm understand correctly it seems like this would wait for every chunk in order. Instead I think we'd want to start running each future (e.g. `map_async`) and `await` them in parallel/poll them all in a loop (but not blocking on any of them), which is what `join_all` seems to do (https://docs.rs/futures-util-preview/0.3.0-alpha.19/src/futures_util/future/join_all.rs.html#115-122)

File: Cargo.toml
Comment (parasyte) [reply Cargo.toml:64]: `async_executor` is used in the example framework: https://github.com/gfx-rs/wgpu-rs/pull/685/files#diff-e3c88eb31749039885929e401e55a90a5f69a0b8cbbcf46cb2df072aafbff069R294

Comment (grovesNL) [reply Cargo.toml:64]: Ah ok, I missed that. Sounds good

File: src/util/belt.rs
Comment (parasyte) [reply src/util/belt.rs]: @grovesNL Glad you caught that. The latest commit rewrites this one to return a future that polls its inner futures similarly to `join_all`.

I've only tested it with the `skybox` demo, but it is properly allowing chunks to be reused in the `StagingBelt`, and I don't see any obvious flaws with the implementation.

File: src/util/belt.rs
Comment (grovesNL) [inline src/util/belt.rs:11]: nit: technically I guess we could just remove items directly from `futures` instead of setting them to `None`, but either seems alright

**Review by grovesNL — APPROVED** [review]
The updated PR looks great, thank you! Could you squash the commits once more before we merge?

File: src/util/belt.rs
Comment (parasyte) [reply src/util/belt.rs:11]: I thought about that, but IIUC, that constitutes a move which would violate `Pin` requirements. Am I wrong?

**Comment by parasyte** [comment]
@grovesNL Thanks for the reviews. This should be ready to go.

**Comment by grovesNL** [comment]
🎉 
bors r+

File: src/util/belt.rs
Comment (grovesNL) [reply src/util/belt.rs:11]: Yeah I think that's right – it would probably have to be behind another pointer type, so maybe not as straightforward as the `Option` approach anyway

**Pull request merged**
